proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+  the rebates is applied. Consider the transaction is executed successfully. The
+  maximum of all the rebates for `accA` is 75 lamports and `accB` is 25
+  lamports. So a total of 100 lamports are rebated back to the payer, `accA`
+  gets 25 lamports and `accB` gets 75 lamports. The payer will end up paying
+  `other fees` + `100` lamports.
+
+* Fees paid full rebates but the execution failed case:
+
+  A payer includes instruction `PayApplicationFees(100)` in the transaction.
+  There is an account `accA` which is write-locked by the transaction and it has
+  an application fee of `100` lamports. Consider during execution the program
+  will rebate all the application fees on the account but later the execution
+  failed. Then payer should have a minimum balance of `other fees` + `100`
+  lamports to execute the transaction. The program rebated application fees but
+  as executing the transaction failed, no rebate will be issued. The application
+  fees will be transferred to respective accounts, and the payer will finally
+  end up paying `other fees` + `100` lamports as application fees.
+
+* Fees are over paid case:
+
+  A payer includes instruction `PayApplicationFees(1000)` in the transaction.
+  There is an account `accA` that is write-locked by the transaction and it has
+  an application fee of `100` lamports. The minimum balance required by payer
+  will be `other fees` + `1000` lamports as application fees. So the payer pays
+  100 lamports for the account as application fees and 900 lamports is an
+  overpayment. The 900 lamports will be transferred back to the user even if the
+  transaction succeeds or fails. The 100 lamports will be transferred to `accA`
+  in all cases except if the transaction is successful and the program issued
+  a rebate.
+
+* Fees underpaid case:
+
+  A payer includes instruction `PayApplicationFees(150)` in the transaction.
+  There is an accounts `accA` that is write-locked by the transaction and it has
+  an application fees of `300` lamports. The minimum balance required by payer
+  will be `other fees` + `150` lamports as application fees to load accounts and
+  schedule transactions for execution. Here payer has insufficiently paid the
+  application fees paying 150 lamports instead of 300 lamports. So before
+  program execution, we detect that the application fees are not sufficiently
+  paid and execution fails with the error `ApplicationFeesNotPaid` and the
+  partially paid amount is transferred back to the payer. So the payer pays only
+  `base fees` in the end but the transaction is unsuccessful.
+
+## Impact
+
+Overall this feature will incentivize the creation of proper transactions and
+spammers would have to pay much more fees reducing congestion in the cluster.
+This will add very low calculation overhead on the validators. It will also
+enable users to protect their accounts against malicious read and write locks.
```

Comment:
> It will also enable users to protect their accounts against malicious read and write locks.

I can see how it protects against write-locks, but not read-locks. I see you mentioned another proposal for read-lock protection, but it's out of the scope of this SIMD, correct?

Reply:
Yes once this is passed we will introduce another one